core/domain/user_services.py

```diff
@@ -2287,3 +2290,60 @@ def get_dashboard_stats(user_id):
         'num_ratings': num_ratings,
         'average_ratings': average_ratings
     }
+
+
+def update_learner_checkpoint_progress(
+    user_id, exploration_id, state_name, exp_version):
+    """Sets the last_completed and latest_visited checkpoint.
+
+    Args:
+        user_id: str. The Id of the user.
+        exploration_id: str. The Id of the exploration.
+        state_name: str. The state name of the last_completed checkpoint.
+        exp_version: int. The exploration version of the last_completed
+            checkpoint.
+    """
+
+    exploration_user_model = user_models.ExplorationUserDataModel.get(
+        user_id, exploration_id)
+    if exploration_user_model is None:
+        exploration_user_model = user_models.ExplorationUserDataModel.create(
+            user_id, exploration_id)
+
+    exploration_user_model.last_completed_checkpoint_exp_version = exp_version
+    # On completing a checkpoint, we update both last_completed and
+    # latest_visited checkpoint.
+    exploration_user_model.last_completed_checkpoint_state_name = state_name
+    exploration_user_model.latest_visited_checkpoint_state_name = state_name
+    exploration_user_model.update_timestamps()
+    exploration_user_model.put()
+
+
+def set_latest_visited_checkpoint(user_id, exploration_id, state_name, version):
+    """Set the last visited checkpoint.
```

Comment:
Docstring does not match function name. Please be really clear about last vs latest since we can't afford to have confusion there.

Reply:
This function has now been removed in the latest changes.